criu/apparmor.c

```diff
@@ -0,0 +1,538 @@
+	 * views, extra //s can indicate a namespace separation. However, I
+	 * think only the apparmor developers use this feature :)
+	 */
+	end = strchr(namespace, ':');
+	if (!end) {
+		pr_err("couldn't find AA namespace end in: %s\n", namespace);
+		return -1;
+	}
+
+	*end = '\0';
+
+	for (i = 0; i < n_namespaces; i++) {
+		/* did we already dump this namespace? */
+		if (!strcmp(namespaces[i]->name, namespace)) {
+			*end = ':';
+			return 0;
+		}
+	}
+
+	pr_info("dumping AA namespace %s\n", namespace);
+
+	ns = new_namespace(namespace, NULL);
+	*end = ':';
+	if (!ns)
+		return -1;
+
+	ret = snprintf(path, sizeof(path), AA_SECURITYFS_PATH "/policy/namespaces/%s", ns->name);
+	if (ret < 0 || ret >= sizeof(path)) {
+		pr_err("snprintf failed?\n");
+		goto err;
+	}
+
+	if (walk_namespace(path, ret, ns) < 0) {
+		pr_err("walking AA namespace %s failed\n", ns->name);
+		goto err;
+	}
+
+	return 0;
+
+err:
+	aa_namespace__free_unpacked(ns, NULL);
+	n_namespaces--;
+	return -1;
+}
+
+int dump_aa_namespaces(void)
+{
+	ApparmorEntry *ae = NULL;
+	int ret;
+
+	if (n_namespaces == 0)
+		return 0;
+
+	ae = xmalloc(sizeof(*ae));
+	if (!ae)
+		return -1;
+	apparmor_entry__init(ae);
+
+	ae->n_namespaces = n_namespaces;
+	ae->namespaces = namespaces;
+
+	ret = pb_write_one(img_from_set(glob_imgset, CR_FD_APPARMOR), ae, PB_APPARMOR);
+
+	apparmor_entry__free_unpacked(ae, NULL);
+	n_namespaces = -1;
+	namespaces = NULL;
+
+	return ret;
+}
+
+bool check_aa_ns_dumping(void)
+{
+	char contents[48];
+	int major, minor, ret;
+	FILE *f;
+
+	f = fopen(AA_SECURITYFS_PATH "/features/domain/stack", "r");
+	if (!f)
+		return false;
+
+	ret = fscanf(f, "%48s", contents);
+	fclose(f);
+	if (ret != 1) {
+		pr_err("scanning aa stack feature failed\n");
+		return false;
+	}
+
+	if (strcmp("yes", contents)) {
+		pr_warn("aa stack featured disabled: %s\n", contents);
+		return false;
+	}
+
+	f = fopen(AA_SECURITYFS_PATH "/features/domain/version", "r");
+	if (!f)
+		return false;
+
+	ret = fscanf(f, "%d.%d", &major, &minor);
+	fclose(f);
+	if (ret != 2) {
+		pr_err("scanning aa stack version failed\n");
+		return false;
+	}
+
+	return major >= 1 && minor >= 2;
+}
+
+static int restore_aa_namespace(AaNamespace *ns, char *path, int offset)
+{
+	pid_t pid;
+	int status;
+
+	pid = fork();
+	if (pid < 0) {
+		pr_perror("fork failed");
+		return -1;
+	}
+
+	if (!pid) {
+		int i, my_offset, ret, fd;
+		char buf[1024];
+
+		ret = snprintf(buf, sizeof(buf), "changeprofile :%s:", ns->name);
+		if (ret < 0 || ret >= sizeof(buf)) {
+			pr_err("profile %s too big\n", ns->name);
+			exit(1);
+		}
+
+		my_offset = snprintf(path+offset, PATH_MAX-offset, "/namespaces/%s", ns->name);
+		if (my_offset < 0 || my_offset >= PATH_MAX-offset) {
+			pr_err("snprintf'd too many characters\n");
+			exit(1);
+		}
+
+		if (mkdir(path, 0755) < 0) {
+			if (errno == EEXIST) {
+				pr_warn("apparmor namespace %s already exists, restoring into it\n", path);
```

Comment:
I'm not sure how unique apparmor namespaces are, but is it possible if the namespace already exists to be used by something else?

Reply:
Yes, this is possible the assumption here is that we are almost always guaranteed to have multiple processes sharing the same namespaces so `write_aa_policy()` would always run into the `EEXIST` condition and it is perfectly fine for different application (independently snapshotted) to share the same AppArmor namespace.